src/Functions/FunctionsJSON.h

```diff
@@ -276,37 +278,139 @@ class FunctionJSONHelpers
+            return result_type->createColumnConstWithDefaultValue(input_rows_count);
+
+        ColumnsWithTypeAndName temporary_columns;
+        if (null_presence.has_nullable)
+            temporary_columns = createBlockWithNestedColumns(arguments);
+        else
+            temporary_columns = arguments;
+
+        ColumnPtr temporary_result;
+
         /// Choose JSONParser.
 #if USE_SIMDJSON
-        if (getContext()->getSettingsRef().allow_simdjson)
-            return FunctionJSONHelpers::Executor<Name, Impl, SimdJSONParser>::run(arguments, result_type, input_rows_count);
+        if (allow_simdjson)
+            temporary_result = FunctionJSONHelpers::Executor<Name, Impl, SimdJSONParser>::run(
+                temporary_columns, result_type, input_rows_count);
+        else
 #endif
-
+        {
 #if USE_RAPIDJSON
-        return FunctionJSONHelpers::Executor<Name, Impl, RapidJSONParser>::run(arguments, result_type, input_rows_count);
+            temporary_result = FunctionJSONHelpers::Executor<Name, Impl, RapidJSONParser>::run(
+                temporary_columns, result_type, input_rows_count);
 #else
-        return FunctionJSONHelpers::Executor<Name, Impl, DummyJSONParser>::run(arguments, result_type, input_rows_count);
+            temporary_result
+                = FunctionJSONHelpers::Executor<Name, Impl, DummyJSONParser>::run(temporary_columns, result_type, input_rows_count);
 #endif
+        }
+
+        if (null_presence.has_nullable)
+            return wrapInNullable(temporary_result, arguments, result_type, input_rows_count);
+        return temporary_result;
+    }
+
+private:
+    NullPresence null_presence;
+    bool allow_simdjson;
+};
+
+
+template <typename Name, template<typename> typename Impl>
+class FunctionBaseFunctionJSON : public IFunctionBase
+{
+public:
+    explicit FunctionBaseFunctionJSON(
+        const NullPresence & null_presence_, bool allow_simdjson_, DataTypes argument_types_, DataTypePtr return_type_)
+        : null_presence(null_presence_)
+        , allow_simdjson(allow_simdjson_)
+        , argument_types(std::move(argument_types_))
+        , return_type(std::move(return_type_))
+    {
+    }
+
+    String getName() const override { return Name::name; }
+
+    const DataTypes & getArgumentTypes() const override
+    {
+        return argument_types;
+    }
+
+    const DataTypePtr & getResultType() const override
+    {
+        return return_type;
+    }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    ExecutableFunctionPtr prepare(const ColumnsWithTypeAndName &) const override
+    {
+        return std::make_unique<ExecutableFunctionJSON<Name, Impl>>(null_presence, allow_simdjson);
+    }
+
+private:
+    NullPresence null_presence;
+    bool allow_simdjson;
+    DataTypes argument_types;
+    DataTypePtr return_type;
+};
+
+
+template <typename Name, template<typename> typename Impl>
```

Comment:
A small comment about why we need to have this `JSONOverloadResolver` and why we cannot just derive from `IFunction` would be useful.

Reply:
Done.